### PR TITLE
[WIP] Add ChatGLM support for vllm

### PR DESCRIPTION
**Pull request by Jeffwan**
Address https://github.com/vllm-project/vllm/issues/231 https://github.com/vllm-project/vllm/issues/247 https://github.com/vllm-project/vllm/issues/607 

I am still new to ChatGLM and model file is still in progress.  I am also looking for collaborators with model experiences for further modifications

### CODE REVIEW

**Comment by Jeffwan** [comment]
@zhuohan123 @WoosukKwon I'd like to use chatglm2 with vLLM together and try my best to contribute here. However, I am not confident on the correctness on the model changes to leverage vllm. Can we collaborate together or there's a better way to reach out to chatglm2 community to give a hand?

File: vllm/model_executor/models/chatglm.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            perform_initialization=False,
+        )
+        self.o_proj = RowParallelLinear(
+            self.projection_size,
+            hidden_size,
+            bias=False,
+            input_is_parallel=True,
+            perform_initialization=False,
+        )
+        self.attn = PagedAttentionWithRoPE(self.num_heads,
```
Comment (irasin) [inline vllm/model_executor/models/chatglm.py:103]: Hi, @Jeffwan, chatglm uses position_encoding_2d in the attention_layer,I think hereis a little problem by just using `PagedAttentionWithRope`. https://huggingface.co/THUDM/chatglm-6b/blob/main/modeling_chatglm.py#L457

Comment (Jeffwan) [reply vllm/model_executor/models/chatglm.py:103]: @irasin Hi Sin, thanks for helping review. Do you like to co-author the PR? I can open the repo write permission to you.

**Comment by Yang-HangWA** [comment]
mark

**Comment by cabbagetalk** [comment]
Why not merge into the main branch and can I use it to support chatglm?

**Comment by liuhuapiaoyuan** [comment]
Mark

File: vllm/model_executor/models/chatglm.py (hunk 1 above)
Comment (Abraham-Xu) [reply vllm/model_executor/models/chatglm.py:103]: Hi, @Jeffwan I do not understand why there is no code about the special attention mask of ChatGLM. Could you please give us some explaination~

**Comment by simon-mo** [comment]
Thank you for the contribution, unfortunately this PR seems to became stale and ChatGLM3 also came out. Feel free to coordinate the contribution here if you have bandwidth! 

https://github.com/vllm-project/vllm/issues/1552

**Comment by zhuohan123** [comment]
Thank you for your contribution! We merged https://github.com/vllm-project/vllm/pull/1261 and the current main branch supports ChatGLM now. Let us know if the main branch does not look good and feel free to propose any changes!

**Pull request closed without merging**